src/mcts/node.h

```diff
@@ -570,6 +553,108 @@ class Edge_Iterator : public EdgeAndNode {
 
+// TODO(crem) Replace this with less hacky iterator once we support C++17.
+// This class has multiple hypostases within one class:
+// * Range (begin() and end() functions)
+// * Iterator (operator++() and operator*())
+// It's more customary to have those as two classes, but
+// creating zoo of classes and copying them around while iterating seems
+// excessive.
+//
+// All functions are not thread safe (must be externally synchronized).
+template <bool is_const>
+class Node_Iterator {
+ public:
+  // Creates "end()" iterator.
+  Node_Iterator() {}
+
+  // Creates "begin()" iterator. Also happens to be a range constructor.
+  // child_ptr will be nullptr if parent_node is solid children.
+  Node_Iterator(const Node& parent_node, Node* child_ptr)
+      : node_ptr_(child_ptr),
+        total_count_(parent_node.num_edges_),
+        solid_(parent_node.solid_children_) {
+    if (node_ptr_ != nullptr && node_ptr_->GetN() == 0) {
+      operator++();
+    }
+  }
+  // These are technically wrong, but are usable to compare with end().
+  bool operator==(const Node_Iterator<is_const>& other) const {
+    return node_ptr_ == other.node_ptr_;
+  }
+  bool operator!=(const Node_Iterator<is_const>& other) const {
+    return node_ptr_ != other.node_ptr_;
+  }
+
+  // Function to support range interface.
+  Node_Iterator<is_const> begin() { return *this; }
+  Node_Iterator<is_const> end() { return {}; }
+
+  // Functions to support iterator interface.
+  // Equality comparison operators are inherited from EdgeAndNode.
+  void operator++() {
+    if (solid_) {
+      while (++current_idx_ != total_count_ &&
+             node_ptr_[current_idx_].GetN() == 0) {
+        if (node_ptr_[current_idx_].GetNInFlight() == 0) {
+          // Once there is not even n in flight, we can skip to the end.
+          current_idx_ = total_count_;
+          break;
+        }
+      }
+      if (current_idx_ == total_count_) {
+        node_ptr_ = nullptr;
+      }
+    } else {
+      do {
+        node_ptr_ = node_ptr_->sibling_.get();
+        // If n started is 0, can jump direct to end.
+        if (node_ptr_ != nullptr && node_ptr_->GetN() == 0 &&
+            node_ptr_->GetNInFlight() == 0) {
+          node_ptr_ = nullptr;
+          break;
+        }
+      } while (node_ptr_ != nullptr && node_ptr_->GetN() == 0);
+    }
+  }
+  Node* operator*() {
+    if (solid_)
+      return &(node_ptr_[current_idx_]);
+    else
+      return node_ptr_;
+  }
+
+ private:
+  // Pointer to current node.
+  Node* node_ptr_ = nullptr;
+  uint16_t current_idx_ = 0;
+  uint16_t total_count_ = 0;
+  bool solid_ = false;
+};
+
+inline Node_Iterator<true> Node::VisitedNodes() const {
+  return {*this, child_.get()};
+}
+inline Node_Iterator<false> Node::VisitedNodes() {
+  return {*this, child_.get()};
+}
+
+inline float Node::GetVisitedPolicy() const {
+  if (n_ == 0 || edges_ == nullptr || child_ == nullptr) return 0.0f;
+  float vp = 0.0f;
+  for (const auto& child : Edges()) {
```

Comment:
Can that new shiny `VisitedNodesIterator` be used here?

Reply:
In 1490 policy caching hasn't yet been removed, so this method doesn't exist - but otherwise yes it could.